src/boot/fwupd.c

```diff
@@ -0,0 +1,630 @@
+
+                capsule_blocks = xrealloc(capsule_blocks,
+                                          n_capsules * sizeof(EFI_CAPSULE_BLOCK_DESCRIPTOR),
+                                          (n_capsules + 1) * sizeof(EFI_CAPSULE_BLOCK_DESCRIPTOR));
+                capsule_blocks[n_capsules] = block_descriptor;
+
+                capsules = xrealloc(capsules,
+                                    n_capsules * sizeof(EFI_CAPSULE_HEADER *),
+                                    (n_capsules + 1) * sizeof(EFI_CAPSULE_HEADER *));
+                capsules[n_capsules++] = TAKE_PTR(capsule);
+        }
+
+        /* Need to terminate the lists with NULL/empty elements as per spec */
+        capsule_blocks = xrealloc(capsule_blocks,
+                                  n_capsules * sizeof(EFI_CAPSULE_BLOCK_DESCRIPTOR),
+                                  (n_capsules + 1) * sizeof(EFI_CAPSULE_BLOCK_DESCRIPTOR));
+        capsule_blocks[n_capsules] = (EFI_CAPSULE_BLOCK_DESCRIPTOR) { };
+        capsules = xrealloc(capsules,
+                            n_capsules * sizeof(EFI_CAPSULE_HEADER *),
+                            (n_capsules + 1) * sizeof(EFI_CAPSULE_HEADER *));
+        capsules[n_capsules] = NULL;
+
+        *ret_capsule_blocks = TAKE_PTR(capsule_blocks);
+        *ret_capsules = TAKE_PTR(capsules);
+        *ret_n_capsules = n_capsules;
+        n_capsules = 0;
+
+        log_debug("Prepared %zu UEFI capsules for update.", *ret_n_capsules);
+
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS apply_updates(
+                EFI_CAPSULE_HEADER **capsules,
+                size_t n_capsules,
+                const EFI_CAPSULE_BLOCK_DESCRIPTOR *capsule_blocks,
+                EFI_RESET_TYPE *ret_reset_type) {
+
+        EFI_RESET_TYPE reset_type;
+        uint64_t max_capsule_size;
+        EFI_STATUS err;
+
+        assert(capsules && n_capsules > 0);
+        assert(capsule_blocks);
+        assert(ret_reset_type);
+
+        err = RT->QueryCapsuleCapabilities(capsules, n_capsules, &max_capsule_size, &reset_type);
+        if (err != EFI_SUCCESS) {
+                log_error_status(err, "Failed to query capsule capabilities, assuming EfiResetWarm: %m");
+                reset_type = EfiResetWarm;
+        }
+
+        err = RT->UpdateCapsule(capsules, n_capsules, POINTER_TO_PHYSICAL_ADDRESS(capsule_blocks));
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to update capsules: %m");
+
+        *ret_reset_type = reset_type;
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS mark_capsules_as_attempted(FwupUpdateTable **updates, size_t n_updates) {
+        EFI_STATUS err;
+
+        FOREACH_ARRAY(u, updates, n_updates) {
+                log_debug("Marking UEFI update variable '%ls' as attempted", (*u)->name);
+
+                EFI_TIME now;
+                err = RT->GetTime(&now, NULL);
+                if (err == EFI_SUCCESS)
+                        (*u)->info->time_attempted = now;
```

Comment:
i am pretty sure if you cannot get the clock just not setting the timestamp is not great, please invalidate the timestamp or just fail altogether